### PR TITLE
Usability/split up pointer primitive checks

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
The pointer-primitive-checks require an address check. Previous we’ve just done these all in in one go, but this makes it hard to see _which_ of our address checks failed.

See https://github.com/diffblue/cbmc/issues/5531
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/cbmc/pointer-primitive-check-01/test.desc
Comment (martin-cs) [inline regression/cbmc/pointer-primitive-check-01/test.desc:1]: Why `broken-smt-backend` when it wasn't before?

File: src/analyses/goto_check.cpp
Comment (martin-cs) [inline src/analyses/goto_check.cpp]: This seems to have less checks?  Thanks for adding the comment but I think I am struggling to understand it, possibly because I am missing the context.  Please can you explain it a bit more to me?

File: src/analyses/goto_check.cpp
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: NIt picking: s/on in/on an/

File: src/analyses/goto_check.cpp
Comment (tautschnig) [reply src/analyses/goto_check.cpp]: Adding to this: is a string comparison really the best solution? Couldn't we instead compare `c.assertion` to `null_object(pointer)`?

**Review by tautschnig — COMMENTED** [review]
As said in my comment below, I'm not convinced the string comparison is the right approach. Apart from this: Please also add a "Fixes: #5531" line to the commit message.

File: regression/cbmc/pointer-primitive-check-01/test.desc
Comment (hannes-steffenhagen-diffblue) [reply regression/cbmc/pointer-primitive-check-01/test.desc:1]: @martin-cs I am not sure _why_ but some individual pointer checks seem to have different results in the smt backend. Because in the previous version of this we AND-ed them all together this didn’t matter before.

File: src/analyses/goto_check.cpp
Comment (hannes-steffenhagen-diffblue) [reply src/analyses/goto_check.cpp]: @tautschnig I don’t like it either, but I don’t like repeating the implementation of `c.assertion` much better. My preferred thing to do would be to add stable IDs to the checks we generate so we can check that, e.g. this is "the address null check", something along those lines. This would be a bigger change though, not sure if people would be OK with that approach?

Comment (tautschnig) [reply src/analyses/goto_check.cpp]: How about adding further members to `conditiont` that track which kinds of properties are already asserted? We don't have to be complete right away, you could just start with a single one: `bool has_null_check`.

Comment (hannes-steffenhagen-diffblue) [reply src/analyses/goto_check.cpp]: @tautschnig One option to get this through right now would be to park the idea of how to filter out checks for later and just accept the redundant null-check for now, but that doesn’t make me happy either.

Comment (tautschnig) [reply src/analyses/goto_check.cpp]: Well, why not try to route of adding, e.g., `has_null_check`?

File: regression/cbmc/pointer-primitive-check-01/test.desc
Comment (martin-cs) [reply regression/cbmc/pointer-primitive-check-01/test.desc:1]: @hannes-steffenhagen-diffblue Thanks for the explanation but this does leave me a little concerned.  They really shouldn't do.  This is kind of important functionality.

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [inline src/analyses/goto_check.cpp]: :no_entry_sign: This doesn't actually appear to be adding the equivalent checks in the case where `mode == ID_java` or where `flags.is_null()` is false.

There must be a way to implement this which is both more readable and more straight forward to get correct. Besides the issue of the string comparison not being great, I think that filtering the list of conditions after generating them is fundamentally a sub-optimal approach. I would suggest starting by looking at splitting the `goto_checkt::address_check` function into multiple functions. At least one of which should generate the list of conditions without the null check. That way the check never has to be filtered out after being unnecessarily added.

File: src/analyses/goto_check.cpp
Comment (hannes-steffenhagen-diffblue) [reply src/analyses/goto_check.cpp]: @tautschnig I don’t completely understand how you’d want that to work – I want to remove the null check, not ask whether it has one.

Comment (tautschnig) [reply src/analyses/goto_check.cpp]: I'd basically use a `bool` member instead of the string. But then what @thomasspriggs [said](https://github.com/diffblue/cbmc/pull/5582/files#r525275170) seems a lot better.

**Comment by tautschnig** [comment]
@hannes-steffenhagen-diffblue Please let us know when this is ready for review - I like the direction this is going, but 1) some cleanup of the previous approach is needed and 2) tests seem to be failing.

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [inline src/analyses/goto_check.cpp]: :no_entry_sign: Adding this when we don't need it is now unnecessary churn. It should be removed so that we are treating all the descriptions in the same way.

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [inline src/analyses/goto_check.cpp]: :question: What do you mean by "fixed"? Is this an assumption that is still being made with the current refactor? Is `pointer_validity_check` making the same assumption?

Comment (thomasspriggs) [reply src/analyses/goto_check.cpp]: :pick: I am not sure the comment about internationalisation is useful. Surely that would affect strings throughout the code base, not just this one. That also seems like the implementation choices for that are something which would be decided when we get to it rather than something which is a useful comment.

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [inline src/analyses/goto_check.cpp:2224]: :bulb: You could just do `const auto &pointer_type = type_checked_cast<pointer_typet>(address.type());` rather than duplicating the same check inside `to_pointer_type` and taking 2 lines to do this.

File: src/analyses/goto_check.cpp
Comment (thomasspriggs) [inline src/analyses/goto_check.cpp:206]: :pick: It would be nice to have some doxygen for these functions.

**Review by thomasspriggs — COMMENTED** [review]
This is looking better. I like that we no longer need the filtering. :+1: . However I think this needs to be split into more commits before merging (and maybe before fixing the tests). This is because we are now doing multiple distinct different things within the same commit. Keeping them in separate commits would help with both debugging the tests right now and making the commit history make sense which could help with tracking down future issues. I think I would make the commit history look something like this -

1. Reorder existing checks in `address_check`. As in - put the null check first only. The tests can then be updated for this functionality only which will help with testing the next commit.
2. Refactor to split the `address_check` function. This commit should include no functionality changes. That way we can use the tests from commit 1 to test that this refactor is sound.
3. Separate calls to `add_guarded_property` in `pointer_primitive_check`. Leave the null condition in place.
4. Remove the null condition in `pointer_primitive_check`. I am assuming you did this removal in order to avoid adding a condition which would otherwise be checking "x is null or x is not null". However the commit history does not currently make this clear and putting this in a separate commit will allow for putting the motivation in the commit message body.

I am a little on the fence about the ordering 3/4.

**Review by thomasspriggs — APPROVED** [review]
Thank you for putting the work into refactoring this and cleaning up the commit history. This version was _much_ easier to review and this implementation should be easier to maintain.

**Review by chrisr-diffblue — APPROVED** [review]
Generally looks good, thanks - I'd recommend getting a re-review from @tautschnig too, just to check he's happy with the updates.

**Comment by hannes-steffenhagen-diffblue** [comment]
@tautschnig are you OK with this now?

**Pull request merged**
